core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -558,97 +558,134 @@ public CreatePathResult createPath(LockedInodePath inodePath, CreatePathOptions<
+          LOG.error(errorMessage);
+          throw new FileAlreadyExistsException(errorMessage);
         }
-        lastInode.setPinned(currentInodeDirectory.isPinned());
-      } else if (options instanceof CreateFileOptions) {
-        CreateFileOptions fileOptions = (CreateFileOptions) options;
-        lastInode = InodeFile.create(mContainerIdGenerator.getNewContainerId(),
-            currentInodeDirectory.getId(), name, System.currentTimeMillis(), fileOptions);
-        // Lock the created inode before subsequent operations, and add it to the lock group.
-        lockList.lockWriteAndCheckNameAndParent(lastInode, currentInodeDirectory, name);
-        if (currentInodeDirectory.isPinned()) {
-          // Update set of pinned file ids.
-          mPinnedInodeFileIds.add(lastInode.getId());
+      } else {
+        // create the new inode, with a write lock
+        if (options instanceof CreateDirectoryOptions) {
+          CreateDirectoryOptions directoryOptions = (CreateDirectoryOptions) options;
+          lastInode = InodeDirectory.create(mDirectoryIdGenerator.getNewDirectoryId(journalContext),
+              currentInodeDirectory.getId(), name, directoryOptions);
+          // Lock the created inode before subsequent operations, and add it to the lock group.
+          lockList.lockWriteAndCheckNameAndParent(lastInode, currentInodeDirectory, name);
+          if (directoryOptions.isPersisted()) {
+            // Do not journal the persist entry, since a creation entry will be journaled instead.
+            InodeUtils.syncPersistDirectory((InodeDirectory) lastInode, this, mMountTable,
+                NoopJournalContext.INSTANCE);
+          }
+        } else if (options instanceof CreateFileOptions) {
+          CreateFileOptions fileOptions = (CreateFileOptions) options;
+          lastInode = InodeFile.create(mContainerIdGenerator.getNewContainerId(),
+              currentInodeDirectory.getId(), name, System.currentTimeMillis(), fileOptions);
+          // Lock the created inode before subsequent operations, and add it to the lock group.
+          lockList.lockWriteAndCheckNameAndParent(lastInode, currentInodeDirectory, name);
+          if (currentInodeDirectory.isPinned()) {
+            // Update set of pinned file ids.
+            mPinnedInodeFileIds.add(lastInode.getId());
+          }
+          if (fileOptions.isCacheable()) {
+            ((InodeFile) lastInode).setCacheable(true);
+          }
         }
         lastInode.setPinned(currentInodeDirectory.isPinned());
-      }
 
-      createdInodes.add(lastInode);
-      mInodes.add(lastInode);
-      currentInodeDirectory.addChild(lastInode);
-      currentInodeDirectory.setLastModificationTimeMs(options.getOperationTimeMs());
-    }
+        if (!currentInodeDirectory.addChild(lastInode)) {
+          // Could not add the child inode to the parent. Continue and try again.
+          lockList.unlockLast();
+          lastInode = null;
+          continue;
+        }
```

Comment:
I updated the code a little bit.

The guideline is to update the state after a successful add, so cleanup is not necessary.